[abandoned] Add native allreduce and handle processgroup as int

wconstab opened the pull request:
This is an older version of this stack:
https://github.com/pytorch/pytorch/pull/93113
Use the newer one.  It's mostly just a cleaned up version of this stack.

ABANDON ME


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92735
* __->__ #92167

**Summary**
- handle processgroup as int
- return subclass for async eager support
- implement a meta function and test it
- support backwards
- support dynamo tracing

**Issues**
- user must import torch.distributed.traceable_collectives to trigger op registration, otherwise no op - need to move this somewhere it can be automatically registered, but ran into import cycles
- python_dispatcher is required for eager; this isn't ideal, I'd rather use python op registration but it doesn't compose with tensor subclasses due to a limitation with torchdeploy
- API isn't complete; proposals include using list of ranks instead of pg_id, and lazily allocating/caching a PG under the hood.  ReduceOp isn't correctly supported
- eager mode perf impact isn't quantified.  how much latency is added by additional dispatches and .clones()?
- Inductor lowering isn't actually right, needs to be smarter about who owns the input buffer (could potentially use mutationlayout as currently, but only for certain inputs; need to allocate a buffer in other cases)
- Something is wrong with the integration of [nccl/c10d/inductor] causing a hang or a crash at process exit sometimes, during ~Work cleanup, need to debug

**Cleanups**
- move op registration
- ProcessGroup registration/cache layer components into one new file with explanation

wconstab committed:
Add native allreduce and handle processgroup as int

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented on allred.py:
```diff
@@ -0,0 +1,56 @@
+    # however, i did it to demonstrate the API proposed, whereby pg as int is passed
+    # to collective APIs and pg object is recovered in execution layer
+    pg = _get_default_group()
+    pg_id = register_process_group(pg)
```
can use pg_name?

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"



Example Inductor IR 
```
def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1 = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf0)
        buf1 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf1)
        buf4 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        buf2 = as_strided(buf4, (4, 4), (4, 1))  # alias
        stream0 = get_cuda_stream(0)
        triton_fused_cat_0.run(buf0, buf2, 16, grid=grid(16), stream=stream0)
        del buf0
        buf3 = as_strided(buf4, (4, 4), (4, 1), 16)  # alias
        triton_fused_cat_0.run(buf1, buf3, 16, grid=grid(16), stream=stream0)
        del buf2
        del buf3
        buf5 = dist.all_reduce(buf4, async_op=True, group=0, op=ReduceOp.SUM)
        buf5.wait()
        del buf5
        buf7 = buf1; del buf1  # reuse
        extern_kernels.mm(arg5_1, arg5_1, out=buf7)
        del arg5_1
        buf8 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        triton_fused_add_1.run(buf4, buf7, buf8, 32, grid=grid(32), stream=stream0)
        return (buf8, )
```

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"



Example Inductor IR 
```
def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1 = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf0)
        buf1 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf1)
        buf4 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        buf2 = as_strided(buf4, (4, 4), (4, 1))  # alias
        stream0 = get_cuda_stream(0)
        triton_fused_cat_0.run(buf0, buf2, 16, grid=grid(16), stream=stream0)
        del buf0
        buf3 = as_strided(buf4, (4, 4), (4, 1), 16)  # alias
        triton_fused_cat_0.run(buf1, buf3, 16, grid=grid(16), stream=stream0)
        del buf2
        del buf3
        buf5 = dist.all_reduce(buf4, async_op=True, group=0, op=ReduceOp.SUM)
        buf5.wait()
        del buf5
        buf7 = buf1; del buf1  # reuse
        extern_kernels.mm(arg5_1, arg5_1, out=buf7)
        del arg5_1
        buf8 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        triton_fused_add_1.run(buf4, buf7, buf8, 32, grid=grid(32), stream=stream0)
        return (buf8, )
```

[ghstack-poisoned]

wconstab commented:
Example Inductor IR 
```
def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1 = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf0)
        buf1 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg5_1, arg5_1, out=buf1)
        buf4 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        buf2 = as_strided(buf4, (4, 4), (4, 1))  # alias
        stream0 = get_cuda_stream(0)
        triton_fused_cat_0.run(buf0, buf2, 16, grid=grid(16), stream=stream0)
        del buf0
        buf3 = as_strided(buf4, (4, 4), (4, 1), 16)  # alias
        triton_fused_cat_0.run(buf1, buf3, 16, grid=grid(16), stream=stream0)
        del buf2
        del buf3
        buf5 = dist.all_reduce(buf4, async_op=True, group=0, op=ReduceOp.SUM)
        buf5.wait()
        del buf5
        buf7 = buf1; del buf1  # reuse
        extern_kernels.mm(arg5_1, arg5_1, out=buf7)
        del arg5_1
        buf8 = empty_strided((8, 4), (4, 1), device='cuda', dtype=torch.float32)
        triton_fused_add_1.run(buf4, buf7, buf8, 32, grid=grid(32), stream=stream0)
        return (buf8, )
```

ezyang commented on allred.py:
```diff
@@ -0,0 +1,52 @@
+import os
```
Probably don't want this here

wconstab replied:
yea, ill axe it when i get closer to landing. it hasn't lost its utility yet

ezyang commented on aten/src/ATen/native/Collectives.cpp:
```diff
@@ -0,0 +1,23 @@
+namespace at {
+namespace native {
+
+// Can we skip having a c++ impl? i will try doing a python_dispatcher impl
```
I think so but we have to impede the codegen from producing the default composite entry. Maybe a explicit but empty dispatch: entry will work?

ezyang commented on torch/distributed/traceable_collectives.py:
```diff
@@ -0,0 +1,92 @@
+        mutated_input, work = c10d.{inplace_collective}(input)
+        return AsyncCollectiveTensor(mutated_input, work)
+    """
+    _tensor: torch.Tensor
```
This is a kind if interesting design as it implies we have to preallocate the tensor while we wait for the Work to resolve. I think that's always true for the distributed collectives?

wconstab replied:
yea, i didn't check the nccl apis but i am only familiar using collectives as an inplace/mutating op.  do you see this as being a problem?

ezyang replied:
No, it should be OK as long as there aren't any situations where you can't afford to preallocate (i.e. memory constrained)

ezyang commented on torch/distributed/traceable_collectives.py:
```diff
@@ -0,0 +1,92 @@
+        # TODO what happens when func is an inplace? (add_) -
+        # currently, seems like the mutation works on the underlying tensor,
+        # but it remains an AsyncCollectiveTensor so subsequent ops still 'unwrap' again.
+        # would be nice to fix this.
```
This won't be so easy to fix as it implies the type of the subclass would have to change. It's probably easiest to make the tensor inert and just return the internal after work is waited on

wconstab replied:
yea, i can do it that way

ezyang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -226,6 +226,25 @@ void _register_builtin_comm_hook(
+    const c10::intrusive_ptr<::c10d::ProcessGroup>& pg) {
+  auto& pg_table = get_process_group_table();
+  int64_t id = pg_table.size();
+  pg_table.push_back(pg);
```
Probably good to setup some thread safety maybe

ezyang reviewed (commented):
I didn't review the test or inductor code. This seems basically reasonable. I'm not sure what you want the merge to master strategy to be

ezyang commented on torch/distributed/traceable_collectives.py:
```diff
@@ -0,0 +1,92 @@
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
```
There are two problems with enabling Python dispatcher by default: performance, and torchdeploy/multiply.

Performance: Python dispatcher has a startup cost (when we do the initial resolution; not a big deal) and then a per op overhead (the cost to do a dictionary lookup to see what the Python dispatcher entry should be). It is cheaper than it used to be, but not free. And of course if you actually hit the Python code path in Python dispatcher, this is a sizable slow down (but you can pay this slow down on a per op and dispatch key basis.)

Torchdeploy/multipy: The reason why Python dispatcher works but Python op registration doesn't is because Python dispatcher is TLS and we assume that the Python dispatcher only needs to be enabled whereever we propagate this TLS. If Python dispatcher is on by default, this TLS is mandatory and we MUST propagate it to all threads that do execution on behalf of the Python interpreter. In a generalized multipy setting this implies changes to multipy code, as multipy C++ code that calls into Python interpreters is not required to keep the Python interpreter on the same thread: it can migrate a Python interpreter to another thread and it must know to propagate the TLS in this case. Maybe there is an easy interposition point to actually do this transferral but when I discussed this with @zdevito and @d4l3k I got convinced to not do it this way (and that's why we have hermetic PyObjects instead.)

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"


**Summary**
- handle processgroup as int
- return subclass for async eager support
- implement a meta function and test it
- support backwards
- support dynamo tracing

**Issues**
- user must import torch.distributed.traceable_collectives to trigger op registration, otherwise no op - need to move this somewhere it can be automatically registered, but ran into import cycles
- python_dispatcher is required for eager; this isn't ideal, I'd rather use python op registration but it doesn't compose with tensor subclasses due to a limitation with torchdeploy
- API isn't complete; proposals include using list of ranks instead of pg_id, and lazily allocating/caching a PG under the hood.  ReduceOp isn't correctly supported
- eager mode perf impact isn't quantified.  how much latency is added by additional dispatches and .clones()?
- Inductor lowering isn't actually right, needs to be smarter about who owns the input buffer (could potentially use mutationlayout as currently, but only for certain inputs; need to allocate a buffer in other cases)
- Something is wrong with the integration of [nccl/c10d/inductor] causing a hang or a crash at process exit sometimes, during ~Work cleanup, need to debug

**Cleanups**
- move op registration
- ProcessGroup registration/cache layer components into one new file with explanation

[ghstack-poisoned]

wconstab committed:
Update on "Add native allreduce and handle processgroup as int"


**Summary**
- handle processgroup as int
- return subclass for async eager support
- implement a meta function and test it
- support backwards
- support dynamo tracing

**Issues**
- user must import torch.distributed.traceable_collectives to trigger op registration, otherwise no op - need to move this somewhere it can be automatically registered, but ran into import cycles
- python_dispatcher is required for eager; this isn't ideal, I'd rather use python op registration but it doesn't compose with tensor subclasses due to a limitation with torchdeploy
- API isn't complete; proposals include using list of ranks instead of pg_id, and lazily allocating/caching a PG under the hood.  ReduceOp isn't correctly supported
- eager mode perf impact isn't quantified.  how much latency is added by additional dispatches and .clones()?
- Inductor lowering isn't actually right, needs to be smarter about who owns the input buffer (could potentially use mutationlayout as currently, but only for certain inputs; need to allocate a buffer in other cases)
- Something is wrong with the integration of [nccl/c10d/inductor] causing a hang or a crash at process exit sometimes, during ~Work cleanup, need to debug

**Cleanups**
- move op registration
- ProcessGroup registration/cache layer components into one new file with explanation

[ghstack-poisoned]
